Improve iteration for multiple documents in one file.

contradict opened the pull request:
The first patch implements some missing functions from the iteration interface to describe the iterator, utilities like `collect` need use some of these.

The second patch fixes `load_all_file` to not close the file before it is done reading it.

I also added a test for each of these.

contradict committed:
Complete iteration protocol.

contradict committed:
Do not close file before iterator reads data.

contradict committed:
Add more iteration tests

Paalon reviewed (commented):
A YAML document can be a single node with a scalar or a sequence, so `Base.eltype` should not return `Dict{Any, Any}`.

contradict commented:
Good point!

I suppose the safest thing to do is change `IteratorEltype` to return `EltypeUnknown()` since a file could switch back and forth between complex documents and single scalar documents.

Is this package now maintained? I'd be happy to keep working on this if there are signs of life!

kescobo commented:
@contradict it is very loosely (and poorly) maintained by me. It's kind of an accident - I didn't write the package, have never dug into the internals, and don't use it day-to-day.

In general, I'm happy to review PRs, and if they provide tests and don't break anything, I'll usually merge and release, though I may need some reminders to pay attention. Sorry this languished so long.

contradict commented:
Ok, rebased and comment addressed. Unfortunately, this is still broken on Windows. I posted [here](https://discourse.julialang.org/t/help-understanding-an-error-on-windows/68059) looking for advice when I made this PR and I haven't made any progress on that part of the problem since.

GunnarFarneback commented on src/YAML.jl:
```diff
@@ -97,8 +100,5 @@ load_file(filename::AbstractString, args...; kwargs...) =
-        load_all(input, args...; kwargs...)
-    end
-
+    load_all(open(filename, "r"), args...; kwargs...)
```
I'm fairly sure this won't close the file, in contrast to the `open(...) do` construction. Is that intentional?

contradict replied:
Yes. Since parsing is lazy, using `do` here results in the file being closed before anything is read.

GunnarFarneback replied:
Ok, but it's still problematic to leave a file unclosed. It's entirely possible that's the cause of the Windows failures and even if it seems to work on the other platforms, it will be a file descriptor leak and if you read too many YAML files this way you will eventually fail because you have run out of file descriptors.

There are some options to handle this, with varying advantages and disadvantages.
1. Design the API with a manually called close operation.
2. Set up a finalizer in `YAMLDocIterator` which closes the file when the iterator is garbage collected.
3. Read the raw file eagerly and do the lazy parsing from an internal `IOBuffer`. That way the file can be closed immediately after reading.

GunnarFarneback replied:
After looking closer at the code I think 3 is the better option. The main drawback is that it requires memory to store the raw file contents but if your file is large enough that that becomes a problem, you should use the `load_all(::IO)` method directly instead.

GunnarFarneback replied:
Good observation that the `IOStream` object would have a finalizer. Unfortunately there are no guarantees that it will run before you run out of file descriptors, so better avoid relying on it.

contradict committed:
Read the whole file into memory

when parsing multi-document files. Since parsing is lazy, just using
`do` results in the file being closed before parsing happens. Leaving
the file open and wating for the finalizer to close it was deemed
undesirable.

GunnarFarneback commented on src/YAML.jl:
```diff
@@ -131,6 +131,9 @@ function iterate(it::YAMLDocIterator, _ = nothing)
     return doc, nothing
 end
 
+Base.IteratorSize(::Type{YAMLDocIterator}) = Base.SizeUnknown()
```
This was implemented (recently) on line 119 but here it looks more correct, or at least more consistent with methods from Base. Can you remove the line 119 method (and probably the comment, unless you think it's worth keeping)?

contradict replied:
Good catch, done.

contradict committed:
Combine duplicated IteratorSize implementation, add comment
